lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1295,25 +1341,88 @@ class MongoClientInterface {
+                // Should not return an error when no object is found
+                if (err.is.NoSuchKey) {
+                    return cb(null);
+                }
                 log.error(
                     'deleteObjectNoVer: error deleting object with no version',
-                    { error: err.message });
+                    { error: err.message, bucket: bucketName, key: objName });
                 return cb(errors.InternalError);
             }
-            if (result.ok !== 1) {
+            if (res.ok !== 1) {
                 log.error(
-                    'deleteObjectNoVer: failed deleting object with no version',
-                    { error: err.message });
+                    'deleteObjectNoVer: failed to delete object',
+                    { bucket: bucketName, object: objName });
                 return cb(errors.InternalError);
             }
             return cb(null);
         });
     }
 
+    /**
+     * Flags the object before deleting it, this is done
+     * to keep object metadata in the oplog, as oplog delete
+     * events don't contain any object metadata
+     * @param {Object} collection MongoDB collection
+     * @param {string} bucketName bucket name
+     * @param {string} key Key of the object to delete
+     * @param {object} filter additional query filters
+     * @param {Logger}log logger instance
+     * @param {Function} cb callback containing error
+     * and BulkWriteResult
+     * @return {undefined}
+     */
+    internalDeleteObject(collection, bucketName, key, filter, log, cb) {
+        // filter used when finding and updating object
+        const findUpdateFilter = Object.assign({
+            _id: key,
+        }, filter);
+        // filter used when deleting object
+        const deleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, filter);
+        async.waterfall([
+            next => collection.findOne(findUpdateFilter, {}, (err, doc) => {
+                if (err) {
+                    log.error('internalDeleteObject: error getting object',
+                        { bucket: bucketName, object: key, error: err.message });
+                    return next(errors.InternalError);
+                }
+                if (!doc) {
+                    log.error('internalDeleteObject: unable to find target object to delete',
+                        { bucket: bucketName, object: key });
+                    return next(errors.NoSuchKey);
+                }
+                const objMetadata = new ObjectMD(doc.value);
+                // adding delete flag to object metadata
+                objMetadata.setDeleted(true);
+                // updating originOp
+                objMetadata.setOriginOp('s3:ObjectRemoved:Delete');
+                return next(null, objMetadata.getValue());
+            }),
+            (objMetadata, next) => collection.bulkWrite([
+                {
+                    updateOne: {
```

Comment:
same here.... though the race condition may be easy to fix:

1. Do a simple update to simply set the delete flag
2. Once it has been marked as deleted, get the full metadata (actually, it may be returned by the first call, if we do a `findAndUpdate(_id: key, delete: false)`)
3. Finalize the delete with a bulkwrite